### PR TITLE
Lament the invincibility of the Turbofish

### PR DESCRIPTION
**Pull request by varkor**
Here a test case is added to ensure that any others attempting to drive the Turbofish to extinction have second thoughts. Previously the [entire test suite would succeed](https://github.com/rust-lang/rust/pull/53511) if generic arguments were accepted without disambiguation, making for [confusing and heartbreaking circumstances](https://github.com/rust-lang/rfcs/pull/2527).

### CODE REVIEW

**Comment by Centril** [comment]
I must say: *"oh noes"*.
Never has there been a sadder day.
Yet for all our woes,
There is great melancholy here, I must say.

File: src/test/ui/bastion-of-the-turbofish.rs
Comment (radix) [inline src/test/ui/bastion-of-the-turbofish.rs]: I'm not sure if there's some sort of poetic device being used here, but to my untrained eye, "a impetuous" seems like an error.

**Comment by llogiq** [comment]
A turbofish, often lamented  
and shunned, even called "demented",  
has the last laugh this time  
so with song and with rhyme  
it is hereby rust-astic cemented.

**Review by sugar700 — APPROVED** [review]

**Comment by wycats** [comment]
@varkor While I definitely agree that this makes it "not a slam dunk", I don't think it means that there is no hope. The case in question is very obscure, would be very unlikely to compile both ways in any real scenario, is easily linted (in the case of ambiguity). It is therefore possible that we could decide to make the change anyway (with a long enough time horizon, and with enough practical surveying), and I wouldn't want someone interested in doing the (harder) work of a transition plan to find this issue and decide that there was no point in even attempting it.

**Comment by varkor** [comment]
@wycats: I personally would love to see a transition plan, though I think the constraints are difficult. The test case here is just to make sure that if someone does intend to tackle the problem, they're aware of this particular ambiguity. I think there's enough context to make clear the point of the test (the origin should be easily located with a quick search).

**Comment by withoutboats** [comment]
Wouldn't propose any change to this PR, but I agree with @wycats that this isn't necessarily *the* hard blocker. I would enumerate these constraints on getting rid of the turbofish:

1. How will we migrate these ambiguous edge cases?
2. How does it impact our grammatical complexity (both formally and informally)? How much more difficult is it to implement a Rust parser?
3. How does the implementation impact compile times?

None of these seem impossible to overcome, but possibly difficult.

**Comment by eddyb** [comment]
We *can* "just" make this test fail in Rust 2018, right? Riiight?

File: src/test/ui/bastion-of-the-turbofish.rs
Comment (aw1621107) [inline src/test/ui/bastion-of-the-turbofish.rs]: // where hope must *be* abandoned?

Comment (varkor) [reply src/test/ui/bastion-of-the-turbofish.rs]: Nicely caught, thanks!

**Comment by xfix** [comment]
Something worth nothing is that in C# 2.0, a breaking change was introduced to change the way `a<b,c>(d)` is parsed. I can see this being feasible to do in Rust as well.

**Comment by mark-i-m** [comment]
@varkor Perhaps the test should include a link to this issue for context?

**Comment by Centril** [comment]
My world is filled with unbounded sorrow.
I feel there is no tomorrow.
All that I had ever wished for,
Burnt down to its core.
But perhaps someone left open the door,
Could there be a solution in store?
For Turbofish to be fixed...
Or will that evermore be nixed?
The opinions seem to be mixed.

**Comment by varkor** [comment]
@mark-i-m: I added one, but slightly too late for the rollup. We can make the change as a follow-up if need be (though given how much attention has been drawn to it now, I imagine it won't be hard to find if anyone stumbles upon it in the test suite 😄).

**Comment by varkor** [comment]
Are things going to go terribly wrong if I re-approve this after it's been merged as a rollup, with the context clarification? Let's see.

@bors r=nagisa rollup

**Pull request merged**
